File: examples/PCR3BP/PCR3BP.py
# PCR3BP.py, https://github.com/bhanson10/gbees/tree/main/examples/PCR3BP
# Copyright 2024 by Benjamin Hanson, published under BSD 3-Clause License.

import sys
sys.path.append('../../')
import gbeespy as gbees  # type: ignore
import math

DIM_f = 4 # State dimension
DIM_h = 3 # Measurement dimension

# This function defines the dynamics model - required
def PCR3BP(x, dx, coef):
    r1 = ((x[0] + coef[0])**2 + (x[1])**2)**1.5
    r2 = ((x[0] - 1 + coef[0])**2 + (x[1])**2)**1.5
    f1 = x[2]
    f2 = x[3]
    f3 = 2*x[3]+x[0]-(coef[0]*(x[0]-1+coef[0])/r2)-((1-coef[0])*(x[0]+coef[0])/r1)
    f4 = -2*x[2]+x[1]-(coef[0]*x[1]/r2)-((1-coef[0])*x[1]/r1)
    return [f1, f2, f3, f4]

# This function defines the measurement model - required if MEASURE == True
def rtrr(x, dx, coef):
    h1 = ((x[0] - (1- coef[0]))**2 + (x[1])**2)**0.5 
    h2 = math.atan2(x[1],  x[0] - (1 - coef[0]))
    h3 = ((x[0] - (1 - coef[0]))*x[2] + x[1]*x[3])/h1
    return [h1, h2, h3]

# This function defines the initial grid boundaries - optional
def PCR3BP_J(x, coef):
    r1 = ((x[0]+coef[0])**2+(x[1])**2)**0.5
    r2 = ((x[0]-1+coef[0])**2+(x[1])**2)**0.5
    J = (x[0])**2.0 + (x[1])**2.0 + (2*(1-coef[0])/r1) + (2*coef[0]/r2) + coef[0]*(1 - coef[0]) - ((x[2])**2.0 + (x[3])**2.0)
    return J

#==================================== Read in initial discrete measurement ==================================#
print("Reading in initial discrete measurement...\n")

P_DIR = "<path_to_pdf>"      # Saved PDFs path
M_DIR = "./measurements"     # Measurement path
M_FILE = "measurement0.txt"  # Measurement file
M = gbees.Meas_create(DIM_f, M_DIR, M_FILE)
#============================================================================================================#

#=========================================== Read in user inputs ============================================#
print("Reading in user inputs...\n")

dx = [None] * DIM_f                             # Grid width, default is half of the std. dev. from the initial measurement 
for i in range(DIM_f):
    dx[i] = (M.cov[i][i]**(0.5))/2
G = gbees.Grid_create(DIM_f, 1E-7, M.mean, dx); # Inputs: (dimension, probability threshold, center, grid width)    
 
coef = [1.901109735892602E-07]                  # PCR3BP trajectory attributes (mu)
T = gbees.Traj_create(len(coef), coef)          # Inputs: (# of coefficients, coefficients)

NUM_DIST = 8                                    # Number of distributions recorded per measurement
NUM_MEAS = 4                                    # Number of measurements
DEL_STEP = 20                                   # Number of steps per deletion procedure
OUTPUT_FREQ = 20                                # Number of steps per output to terminal
OUTPUT = False                                  # Write info to terminal
RECORD = False                                  # Write PDFs to .txt file
MEASURE = True                                  # Take discrete measurement updates
BOUNDS = True                                   # Add inadmissible regions to grid
#============================================================================================================#

#================================================== GBEES ===================================================#
gbees.run_gbees(PCR3BP, rtrr, PCR3BP_J, G, M, T, P_DIR, M_DIR, NUM_DIST, NUM_MEAS, DEL_STEP, OUTPUT_FREQ, DIM_h, OUTPUT, RECORD, MEASURE, BOUNDS)

File: examples/PCR3BP/PCR3BP.c
// PCR3BP.c, https://github.com/bhanson10/gbees/tree/main/examples/PCR3BP
// Copyright 2024 by Benjamin Hanson, published under BSD 3-Clause License.

#include "../../gbees.h" // REF- do not include c files
#include "PCR3BP.h"

// This function defines the dynamics model - required
void PCR3BP(double* f, double* x, double* dx, double* coef){
    double r1 = pow(pow(x[0]+coef[0],2)+pow(x[1],2),1.5);
    double r2 = pow(pow(x[0]-1+coef[0],2)+pow(x[1],2),1.5);
    f[0] = x[2];
    f[1] = x[3];
    f[2] = 2*x[3]+x[0]-(coef[0]*(x[0]-1+coef[0])/r2)-((1-coef[0])*(x[0]+coef[0])/r1);
    f[3] = -2*x[2]+x[1]-(coef[0]*x[1]/r2)-((1-coef[0])*x[1]/r1);
}

// This function defines the measurement model - required if MEASURE == true
void rtrr(double* h, double* x, double* dx, double* coef){
    h[0] = pow(pow(x[0] - (1 - coef[0]), 2) + pow(x[1], 2),0.5); 
    h[1] = atan2(x[1], x[0] - (1 - coef[0])); 
    h[2] = ((x[0] - (1 - coef[0]))*x[2] + x[1]*x[3])/h[0];
}

// This function defines the initial grid boundaries - optional
double PCR3BP_J(double* x, double* coef){
    double r1 = pow(pow(x[0]+coef[0],2)+pow(x[1],2), 0.5);
    double r2 = pow(pow(x[0]-1+coef[0],2)+pow(x[1],2), 0.5);
    double J = pow(x[0], 2.0) + pow(x[1], 2.0) + (2*(1-coef[0])/r1) + (2*coef[0]/r2) + coef[0]*(1 - coef[0]) - (pow(x[2], 2.0) + pow(x[3], 2.0));
    return J;
}

int main(void){
    //=================================== Read in initial discrete measurement =================================//
    printf("Reading in initial discrete measurement...\n\n");

    char* P_DIR = "./results/c"; // Saved PDFs path
    char* M_DIR = "./measurements";    // Measurement path
    char* M_FILE = "measurement0.txt"; // Measurement file
    Meas M = Meas_create(DIM_f, M_DIR, M_FILE);
    //==========================================================================================================//

    //========================================== Read in user inputs ===========================================//
    printf("Reading in user inputs...\n\n");
    
    double del[DIM_f];                              // Grid width, default is half of the std. dev. from the initial measurement 
    for(int i = 0; i < DIM_f; i ++){
        del[i] = pow(M.cov[i][i],0.5)/2;
    }
    Grid G = Grid_create(DIM_f, 1E-7, M.mean, del); // Inputs: (dimension, probability threshold, center, grid width)       

    double coef[] = {1.901109735892602E-07};        // PCR3BP trajectory attributes (mu)
    Traj T = Traj_create(1, coef);                  // Inputs: (# of coefficients, coefficients)

    int NUM_DIST = 8;                               // Number of distributions recorded per measurement
    int NUM_MEAS = 4;                               // Number of measurements
    int DEL_STEP = 20;                              // Number of steps per deletion procedure
    int OUTPUT_FREQ = 20;                           // Number of steps per output to terminal
    bool OUTPUT = false;                            // Write info to terminal
    bool RECORD = true;                             // Write PDFs to .txt file
    bool MEASURE = true;                            // Take discrete measurement updates
    bool BOUNDS = true;                             // Add inadmissible regions to grid
    //==========================================================================================================//

    //================================================= GBEES ==================================================//
    run_gbees(PCR3BP, rtrr, PCR3BP_J, G, M, T, P_DIR, M_DIR, NUM_DIST, NUM_MEAS, DEL_STEP, OUTPUT_FREQ, DIM_h, OUTPUT, RECORD, MEASURE, BOUNDS);

    return 0;
}